### PR TITLE
Better errors for Debug and Display traits

### LINKED ISSUE #85844: Reduce verbosity of E0277 for `Debug` and `Display`
On missing `Debug` implementation, the current output is unnecessarily verbose:

```
error[E0277]: `Group` doesn't implement `Debug`
  --> src/tests/a.rs:19:5
   |
19 |     assert_eq!(result.unwrap(), &group);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Group` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `Group`
   = note: add `#[derive(Debug)]` or manually implement `Debug`
   = note: required because of the requirements on the impl of `Debug` for `&Group`
   = note: 1 redundant requirements hidden
   = note: required because of the requirements on the impl of `Debug` for `&&Group`
   = note: required by `std::fmt::Debug::fmt`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

We should special case this and `Display` to give less overwhelming output:

```
error[E0277]: `Group` doesn't implement `Debug`
  --> src/tests/a.rs:19:5
   |
19 |     assert_eq!(result.unwrap(), &group);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Group` cannot be formatted using `{:?}`
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: add `#[derive(Debug)]` to `Group` or manually `impl Debug for Group`
  |
X | #[derive(Debug)] pub struct Group {
  |
```

_Taken from https://blog.frankel.ch/start-rust/1/_

### PR DESCRIPTION
**Pull request by inquisitivecrystal**
Currently, if someone tries to pass value that does not implement `Debug` or `Display` to a formatting macro, they get a very verbose and confusing error message. This PR changes the error messages for missing `Debug` and `Display` impls to be less overwhelming in this case, as suggested by #85844. I was a little less aggressive in changing the error message than that issue proposed. Still, this implementation would be enough to reduce the number of messages to be much more manageable. 

After this PR, information on the cause of an error involving a `Debug` or `Display` implementation would suppressed if the requirement originated within a standard library macro. My reasoning was that errors originating from within a macro are confusing when they mention details that the programmer can't see, and this is particularly problematic for `Debug` and `Display`, which are most often used via macros. It is possible that either a broader or a narrower criterion would be better. I'm quite open to any feedback.

Fixes #85844.

### CODE REVIEW

File: library/core/src/fmt/mod.rs [hunk 1]
```diff
@@ -564,7 +564,7 @@ impl Display for Arguments<'_> {
     on(
         crate_local,
         label = "`{Self}` cannot be formatted using `{{:?}}`",
-        note = "add `#[derive(Debug)]` or manually implement `{Debug}`"
+        note = "add `#[derive(Debug)]` to `{Self}` or manually implement `{Debug}` for `{Self}`"
```
Comment (JohnTitor) [inline library/core/src/fmt/mod.rs:567]: This sounds a bit verbose, I like @estebank's wording:

```suggestion
        note = "add `#[derive(Debug)]` or manually `impl {Debug} for {Self}`"
```

Comment (inquisitivecrystal) [reply library/core/src/fmt/mod.rs:567]: @JohnTitor How would you feel about keeping the to Self in the first half, and only making the change in the second half? I'm a bit concerned that having to read to the end to get the `for {Self}` component may make the message less readable.

Comment (JohnTitor) [reply library/core/src/fmt/mod.rs:567]: Hmm yeah, makes sense. It's fine to me.

Comment (inquisitivecrystal) [reply library/core/src/fmt/mod.rs:567]: Thanks!

**Pull request merged**